spec/models/screened_ip_address_spec.rb

```diff
@@ -320,4 +320,63 @@ def test_bad_value(arg)
       end
     end
   end
+
+  describe '#roll_up' do
+    it 'rolls up IPv4 addresses' do
+      # roll up to /24 requires 256 * 1% ~ 3 IPs
+      # roll up to /22 requires 1024 * 1% ~ 10 IPs
+      SiteSetting.min_ban_entries_for_roll_up = 1
+
+      Fabricate(:screened_ip_address, ip_address: "1.1.1.1")
+      expect { ScreenedIpAddress.roll_up }.not_to change { ScreenedIpAddress.count }
+
+      Fabricate(:screened_ip_address, ip_address: "1.1.1.2")
+      expect { ScreenedIpAddress.roll_up }.not_to change { ScreenedIpAddress.count }
+
+      Fabricate(:screened_ip_address, ip_address: "1.1.1.3")
+      expect { ScreenedIpAddress.roll_up }.to change { ScreenedIpAddress.count }.by(-2)
+      expect(ScreenedIpAddress.pluck(:ip_address)).to include("1.1.1.0/24")
+      expect(ScreenedIpAddress.pluck(:ip_address)).not_to include("1.1.1.1", "1.1.1.2", "1.1.1.3")
+
+      # roll up is stable
+      expect { ScreenedIpAddress.roll_up }.to change { ScreenedIpAddress.count }.by(0)
+      expect(ScreenedIpAddress.pluck(:ip_address)).to include("1.1.1.0/24")
+      expect(ScreenedIpAddress.pluck(:ip_address)).not_to include("1.1.1.1", "1.1.1.2", "1.1.1.3")
+    end
+
+    it 'rolls up IPv6 addresses' do
+      # roll up to /64 requires 256 * 1% ~ 3 IPs
+      # roll up to /60 requires 4096 * 1% ~ 40 IPs
+      SiteSetting.min_ban_entries_for_roll_up = 1
+
+      Fabricate(:screened_ip_address, ip_address: "2001:db8:3333:4441:5555:6666:7777:8888")
+      expect { ScreenedIpAddress.roll_up }.not_to change { ScreenedIpAddress.count }
+
+      Fabricate(:screened_ip_address, ip_address: "2001:db8:3333:4441:5555:6666:7777:8889")
+      expect { ScreenedIpAddress.roll_up }.not_to change { ScreenedIpAddress.count }
+
+      Fabricate(:screened_ip_address, ip_address: "2001:db8:3333:4441:5555:6666:7777:888a")
+      expect { ScreenedIpAddress.roll_up }.to change { ScreenedIpAddress.count }.by(-2)
+      expect(ScreenedIpAddress.pluck(:ip_address)).to include("2001:db8:3333:4441:5555:6666:7777:8888/64")
```

Comment:
this should be `2001:db8:3333:4441::/64`

Reply:
This brings to mind… why did it work?

There's a bit of a quirk here around `include`:

```
      expect(ScreenedIpAddress.pluck(:ip_address)).to include("2001:db8:3333:4441::/65")
```
would also pass here, despite *that particular subnet* not actually being in the array

```
[46] pry(main)> [IPAddr.new('2001:db8::/32')].include? IPAddr.new('2001:db8::/128')
=> true
[47] pry(main)> [IPAddr.new('2001:db8::/32')].include? IPAddr.new('2001:db8::1/128')
=> false
[48] pry(main)> IPAddr.new('2001:db8::/32').include? IPAddr.new('2001:db8::1/128')
=> true
```

```
[52] pry(main)> IPAddr.new('2001:db8::/32') == IPAddr.new('2001:db8::/128')
=> true
```
🤦🏻 it's comparing strings

```
[53] pry(main)> IPAddr.new('2001:db8::/32').eql? IPAddr.new('2001:db8::/128')
=> false
```